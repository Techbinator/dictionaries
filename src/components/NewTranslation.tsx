import React from "react";
import { TaddTranslation, ITranslations } from "../store/Dictionaries.type";
import { validateTranslation } from "../helpers/validation";
interface INewTranslation {
  addTranslation: TaddTranslation;
  dictionaryUUID: string;
  translations: ITranslations;
  translationsUUID: string[];
}

const INITIAL_STATE = {
  domain: "",
  range: "",
  domainError: "",
  rangeError: "",
  severity: ""
};
export default class NewTranslation extends React.Component<INewTranslation> {
  state = INITIAL_STATE;
  onSubmit = (e: React.MouseEvent<HTMLElement>) => {
    const { addTranslation, dictionaryUUID } = this.props;

    const { domain, range } = this.state;

    addTranslation(dictionaryUUID, {
      uuid: "",
      domain,
      range
    });

    this.setState({ ...INITIAL_STATE });
  };
  onInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { translationsUUID, translations } = this.props;

    const inputName = e.target.name as "range" | "domain";

    const { domain, range } = this.state;
    const validation = validateTranslation(
      inputName == "domain" ? e.target.value : domain,
      inputName == "range" ? e.target.value : range,
      translationsUUID,
      translations
    );
    this.setState({ ...validation, [inputName]: e.target.value });
  };
  render() {
    const { domain, range, domainError, rangeError, severity } = this.state;
    return (
      <tr>
        <td style={{ width: "40%" }}>
          <input
            value={domain}
            name="domain"
            onChange={this.onInputChange}
            type="text"
            placeholder="Domain"
          />
          {domainError && <span className={severity}>{domainError}</span>}
        </td>
        <td style={{ width: "40%" }}>
          <input
            value={range}
            onChange={this.onInputChange}
            name="range"
            type="text"
            placeholder="Range"
          />
          {rangeError && <span className={severity}>{rangeError}</span>}
        </td>
        <td style={{ width: "20%" }}>
          <input
            type="button"
            onClick={this.onSubmit}
            value="Add translation"
          />
        </td>
      </tr>
    );
  }
}
