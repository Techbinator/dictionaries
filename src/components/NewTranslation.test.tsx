import React from "react";
import NewTranslation from "./NewTranslation";
import { shallow } from "enzyme";
import contextMock from "../helpers/__mocks__/context.mock";

describe("NewTranslation component", () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallow(
      <NewTranslation
        addTranslation={contextMock.addTranslation}
        dictionaryUUID="619ce930-4116-11e9-9306-ef359c18301d"
        translations={contextMock.translations}
        translationsUUID={
          contextMock.dictionaries["619ce930-4116-11e9-9306-ef359c18301d"]
            .translationsUUID
        }
      />
    );
  });
  it("should render snapshot with default values", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should add translation", () => {
    wrapper.find('input[name="domain"]').simulate("change", {
      target: { value: "Domain", name: "domain" }
    });
    wrapper.find('input[name="range"]').simulate("change", {
      target: { value: "Range", name: "range" }
    });
    wrapper.find('input[type="button"]').simulate("click");
    expect(contextMock.addTranslation).toBeCalledWith(
      "619ce930-4116-11e9-9306-ef359c18301d",
      {
        uuid: "",
        domain: "Domain",
        range: "Range"
      }
    );
  });

  it("should throw validation error", () => {
    wrapper.find('input[name="domain"]').simulate("change", {
      target: { value: "Domain", name: "domain" }
    });
    wrapper.find('input[name="range"]').simulate("change", {
      target: { value: "Domain", name: "range" }
    });
    expect(
      wrapper
        .find(".error")
        .first()
        .text()
    ).toEqual("Domain must not equal range");
  });
});
