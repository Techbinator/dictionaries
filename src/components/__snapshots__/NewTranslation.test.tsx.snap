// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTranslation component should render snapshot with default values 1`] = `
<tr>
  <td
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <input
      name="domain"
      onChange={[Function]}
      placeholder="Domain"
      type="text"
      value=""
    />
  </td>
  <td
    style={
      Object {
        "width": "40%",
      }
    }
  >
    <input
      name="range"
      onChange={[Function]}
      placeholder="Range"
      type="text"
      value=""
    />
  </td>
  <td
    style={
      Object {
        "width": "20%",
      }
    }
  >
    <input
      onClick={[Function]}
      type="button"
      value="Add translation"
    />
  </td>
</tr>
`;
